cpp/src/arrow/compute/cast.cc

```diff
@@ -226,19 +254,237 @@ struct CastFunctor<O, I,
+  INIT_BITSET(valid_bits, indices.offset());
+
+  const index_c_type* in =
+      reinterpret_cast<const index_c_type*>(indices.data()->buffers[1]->data()) +
+      indices.offset();
+  uint8_t* out = output->buffers[1]->mutable_data();
+  int32_t byte_width =
+      static_cast<const FixedSizeBinaryType&>(*output->type).byte_width();
+  for (int64_t i = 0; i < indices.length(); ++i) {
+    if (bitset_valid_bits & (1 << bit_offset_valid_bits)) {
+      const uint8_t* value = dictionary.Value(in[i]);
+      memcpy(out + i * byte_width, value, byte_width);
+    }
+    READ_NEXT_BITSET(valid_bits);
+  }
+}
+
+template <typename T>
+struct CastFunctor<
+    T, DictionaryType,
+    typename std::enable_if<std::is_base_of<FixedSizeBinaryType, T>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options, const Array& input,
+                  ArrayData* output) {
+    const DictionaryArray& dict_array = static_cast<const DictionaryArray&>(input);
+    const DictionaryType& type = static_cast<const DictionaryType&>(*input.type());
+    const DataType& values_type = *type.dictionary()->type();
+    const FixedSizeBinaryArray& dictionary =
+        static_cast<const FixedSizeBinaryArray&>(*type.dictionary());
+
+    // Check if values and output type match
+    bool types_equal;
+    FUNC_RETURN_NOT_OK(TypeEquals(values_type, *output->type, &types_equal));
+    if (!types_equal) {
+      std::stringstream ss;
+      ss << "Dictionay and target type do not match: ";
+      ss << values_type.ToString() << ", " << output->type->ToString();
+      ctx->SetStatus(Status::Invalid(ss.str()));
+      return;
+    }
+
+    const Array& indices = *dict_array.indices();
+    switch (indices.type()->id()) {
+      case Type::INT8:
+        UnpackFixedSizeBinaryDictionary<Int8Type>(ctx, indices, dictionary, output);
+        break;
+      case Type::INT16:
+        UnpackFixedSizeBinaryDictionary<Int16Type>(ctx, indices, dictionary, output);
+        break;
+      case Type::INT32:
+        UnpackFixedSizeBinaryDictionary<Int32Type>(ctx, indices, dictionary, output);
+        break;
+      case Type::INT64:
+        UnpackFixedSizeBinaryDictionary<Int64Type>(ctx, indices, dictionary, output);
+        break;
+      default:
+        std::stringstream ss;
+        ss << "Invalid index type: " << indices.type()->ToString();
+        ctx->SetStatus(Status::Invalid(ss.str()));
+        return;
+    }
+  }
+};
+
+template <typename IndexType>
+Status UnpackBinaryDictionary(FunctionContext* ctx, const Array& indices,
+                              const BinaryArray& dictionary, ArrayData* output) {
+  using index_c_type = typename IndexType::c_type;
+  std::unique_ptr<ArrayBuilder> builder;
+  RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), output->type, &builder));
+  BinaryBuilder& binary_builder = static_cast<BinaryBuilder&>(*builder);
+
+  const uint8_t* valid_bits = indices.null_bitmap_data();
+  INIT_BITSET(valid_bits, indices.offset());
```

Comment:
We should create a helper functor to help with bitmap checking, I will open an issue